omaha/protocol.go

```diff
@@ -0,0 +1,263 @@
+
+	// update engine extension, duplicates the version attribute.
+	UpdaterVersion string `xml:"updaterversion,attr,omitempty"`
+}
+
+func NewRequest() *Request {
+	return &Request{
+		Protocol: "3.0",
+		Version:  version.Version,
+		OS: &OS{
+			Platform: LocalPlatform(),
+			Arch:     LocalArch(),
+			// TODO(marineam): Version and ServicePack
+		},
+	}
+}
+
+func (r *Request) AddApp(id, version string) *AppRequest {
+	a := &AppRequest{Id: id, Version: version}
+	r.Apps = append(r.Apps, a)
+	return a
+}
+
+type AppRequest struct {
+	Ping        *PingRequest    `xml:"ping"`
+	UpdateCheck *UpdateRequest  `xml:"updatecheck"`
+	Events      []*EventRequest `xml:"event" json:",omitempty"`
+	Id          string          `xml:"appid,attr,omitempty"`
+	Version     string          `xml:"version,attr,omitempty"`
+	NextVersion string          `xml:"nextversion,attr,omitempty"`
+	Lang        string          `xml:"lang,attr,omitempty"`
+	Client      string          `xml:"client,attr,omitempty"`
+	InstallAge  string          `xml:"installage,attr,omitempty"`
+
+	// update engine extensions
+	Track     string `xml:"track,attr,omitempty"`
+	FromTrack string `xml:"from_track,attr,omitempty"`
+	Board     string `xml:"board,attr,omitempty"`
+	DeltaOK   bool   `xml:"delta_okay,attr,omitempty"`
+
+	// coreos update engine extensions
+	BootId       string `xml:"bootid,attr,omitempty"`
+	MachineID    string `xml:"machineid,attr,omitempty"`
+	OEM          string `xml:"oem,attr,omitempty"`
+	OEMVersion   string `xml:"oemversion,attr,omitempty"`
+	AlephVersion string `xml:"alephversion,attr,omitempty"`
+}
+
+func (a *AppRequest) AddUpdateCheck() *UpdateRequest {
+	a.UpdateCheck = &UpdateRequest{}
+	return a.UpdateCheck
+}
+
+func (a *AppRequest) AddPing() *PingRequest {
+	a.Ping = &PingRequest{Active: 1}
+	return a.Ping
+}
+
+func (a *AppRequest) AddEvent() *EventRequest {
+	event := &EventRequest{}
+	a.Events = append(a.Events, event)
+	return event
+}
+
+type UpdateRequest struct {
+	TargetVersionPrefix string `xml:"targetversionprefix,attr,omitempty"`
+}
+
+type PingRequest struct {
+	Active               int `xml:"active,attr,omitempty"`
+	LastActiveReportDays int `xml:"a,attr,omitempty"`
+	LastReportDays       int `xml:"r,attr,omitempty"`
+}
+
+type EventRequest struct {
+	Type            EventType   `xml:"eventtype,attr"`
+	Result          EventResult `xml:"eventresult,attr"`
+	NextVersion     string      `xml:"nextversion,attr,omitempty"`
+	PreviousVersion string      `xml:"previousversion,attr,omitempty"`
+	ErrorCode       string      `xml:"errorcode,attr,omitempty"`
+}
+
+// Response sent by the Omaha server
+type Response struct {
+	XMLName  xml.Name       `xml:"response" json:"-"`
+	DayStart DayStart       `xml:"daystart"`
+	Apps     []*AppResponse `xml:"app"`
+	Protocol string         `xml:"protocol,attr"`
+	Server   string         `xml:"server,attr"`
+}
+
+func NewResponse() *Response {
+	return &Response{
+		Protocol: "3.0",
+		Server:   "mantle",
```

Comment:
Fixed, and I would say it is configurable in the sense that the server can optionally set the attribute itself after calling NewResponse.